### ISSUE TITLE
Impossible to make a working module RPM?

### ISSUE THREAD

**Issue by swelljoe** [webmin/webmin#1267]
I'm not sure what changed or when, but makemodulerpm.pl (and makemoduledeb.pl to I think a lesser degree) can't possibly make a working package.

The problem is that when building, you can choose `--rpm-depends` or not...if you choose `--rpm-depends` it constructs a list of all dependencies including those provided by Webmin package itself, but Webmin package apparently doesn't explicitly say it provides. e.g. wbm-cron. So, when installing the package yum/dnf/rpm will detect wbm-cron as missing and refuse to install the package.

But, if `--rpm-depends` is not used (this used to work, not sure why it doesn't now), the post-install dependency checker is run and it will fail if the non-Webmin core dependencies are not already installed. e.g. when installing any Virtualmin module (wbm-virtualmin-awstats) as part of the virtualmin-lamp-stack yum group (https://github.com/virtualmin/virtualmin-yum-groups/blob/master/virtualmin-centos-8-comps.xml) Virtualmin won't yet be installed, and RPM won't know that it needs to be. So, the install fails with a prein error.

Possible solutions:

 - Make Webmin "Provide" all of its core modules. This will allow `--rpm-depends` to work, as-is, I think. I think this is also how it was intended to work, maybe?
 - Make makemodulerpm.pl skip Webmin core modules when calculating dependencies and just list the non-core modules. This will work for our needs and doesn't make a crazy long list of Provides (which feels sub-optimal and messy), but if Webmin minimal is the base package being used it will not work (back to the prein problem, where the package manager won't know about a dependency but Webmin will and will fail during scripts). But, if we depend on `webmin` and webmin minimal package is called `webmin-minimal` this would also be fine.
 - Make all rpm packages skip the postinstall test for dependencies. This is, I think, how it used to work. This would make our Virtualmin packages work with minimal changes to upstream Webmin.

I *think* the Debian package already gets the --deb-depends thing right. Though I need to test with deb-depends, but I'm pretty sure I see Provides for Webmin core modules in the Webmin package.

**Comment by jcameron** [comment]
I think skipping Webmin core modules is the reasonable answer ... but I'll take a look.

**Comment by jcameron** [comment]
Actually, that should already happen, assuming `makemodulerpm.pl` is run in the Webmin base directory (which is possibly a bad assumption?)

**Comment by swelljoe** [comment]
Aha! That's it. It does not run in a Webmin base directory on the build system. It does a clean check out from git into it's own directory all by itself. I kinda thought that'd be better from a reproducibility and testing and security perspective...but, it explains why it used to work. I used to build in a Webmin dir.

**Comment by jcameron** [comment]
Sorry about that - I have checked in an update to makemodulerpm.pl that validates that the expected module files exist.

**Comment by swelljoe** [comment]
This seems worse?

Now it cannot build without all of the extra modules existing in the webmin directory, but we need the extra modules to be dependencies in the package. i.e. when I build `virtualmin-nginx` it depends on `virtual-server`...if it exists in the build webmin directory, it gets excluded (I think?) from the deps list. If it doesn't exist, makemodulerpm.pl exits. The problem we need to solve is packages that can be installed without the package manager being aware of the dependencies, so we can't end up with the postinstall check failing.

I think I just need to make sure the build always happens in a Webmin directory? (And back this change out.)

**Comment by swelljoe** [comment]
Yeah, confirmed. This should be rolled back. It makes it impossible to get the right dependencies in the package.

It was a misunderstanding on my part...an assumption that because I got packages that mostly worked when built outside of the webmin dir, that it was reasonable to do that. I'll finish converting my build scripts to always build in a Webmin directory and clean up after each build (so we get the full list of dependencies).

**Comment by iliarostovtsev** [comment]
This is the patch you need.

https://github.com/webmin/webmin/commit/c56591bd63d031c6cae06b5a258938b797722b91
